presto-spi/src/main/java/com/facebook/presto/spi/block/BlockFlattener.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.api.Experimental;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.spi.block.ClosingBlockLease.newLease;
+import static java.util.Objects.requireNonNull;
+
+@Experimental
+public class BlockFlattener
+{
+    private final ArrayAllocator allocator;
+
+    public BlockFlattener(ArrayAllocator allocator)
+    {
+        this.allocator = requireNonNull(allocator, "allocator is null");
+    }
+
+    public BlockLease flatten(Block block)
+    {
+        requireNonNull(block, "block is null");
+        if (block instanceof DictionaryBlock) {
+            return flattenDictionary((DictionaryBlock) block);
+        }
+        if (block instanceof RunLengthEncodedBlock) {
+            return flattenRLE((RunLengthEncodedBlock) block);
+        }
+        return newLease(block);
+    }
+
+    private BlockLease flattenDictionary(DictionaryBlock dictionaryBlock)
+    {
+        Block block = dictionaryBlock.getDictionary();
+        int[] remappedIds = (int[]) dictionaryBlock.getIds().getBase();
+        int positionCount = dictionaryBlock.getPositionCount();
+        int[] leasedIds = null;
+        while (true) {
+            if (block instanceof DictionaryBlock) {
+                dictionaryBlock = (DictionaryBlock) block;
+                int[] ids = (int[]) dictionaryBlock.getIds().getBase();
+
+                if (leasedIds == null) {
+                    leasedIds = allocator.borrowIntArray(positionCount);
+                }
+
+                for (int i = 0; i < positionCount; ++i) {
+                    leasedIds[i] = ids[remappedIds[i] + dictionaryBlock.getOffsetBase()];
+                }
+
+                remappedIds = leasedIds;
+                block = dictionaryBlock.getDictionary();
+            }
+            else if (block instanceof RunLengthEncodedBlock) {
+                RunLengthEncodedBlock rle = (RunLengthEncodedBlock) block;
+                int[] ids = allocator.borrowIntArray(positionCount);
```

Comment:
It seems to me that `ids` array is not necessary and this block could be simplified to 

```
                RunLengthEncodedBlock rle = (RunLengthEncodedBlock) block;
                if (leasedIds == null) {
                    leasedIds = allocator.borrowIntArray(positionCount);
                }
                Arrays.fill(leasedIds, 0, positionCount, 0);
                remappedIds = leasedIds;

                block = rle.getValue();
```

Reply:
Good observation, thanks